drivers: flash: stm32 spi flash driver compile with MISRA_SANE

FRASTM opened the pull request:
When compiling the drivers/flash for stm32 devices with CONFIG_MISRA_SANE=y, the raw[] table must have a defined size and not variable one. Fix it with decl_nph = 2.

Fixes https://github.com/zephyrproject-rtos/zephyr/issues/65795

erwango commented on drivers/flash/flash_stm32_ospi.c:
```diff
@@ -2168,9 +2168,17 @@ static int flash_stm32_ospi_init(const struct device *dev)
+ * with decl_nph = 2
+ * else ISO C90 forbids array 'raw' whose size cannot be evaluated
+ */
+#define JESD216_SFDP_SIZE_NPH (sizeof(struct jesd216_sfdp_header)	\
+				+ ((2) * sizeof(struct jesd216_param_header)))
```
Can't we move definition in `drivers/flash/jesd216.h`, and replace faulty `JESD216_SFDP_SIZE()` ?

de-nordic replied:
I think that would be the right way to go.

FRASTM committed:
drivers: flash: stm32 spi flash driver compile with MISRA_SANE

When compiling the drivers/flash for stm32 devices with
CONFIG_MISRA_SANE=y, the raw[] table must have a defined size
and not variable one. Fix it with decl_nph = 2.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

erwango commented:
> @de-nordic, @erwango Do you mean that decl_nph = 2 whatever CONFIG_MISRA_SANE, so that JESD216_SFDP_SIZE is no more a function of nph ?

Yes

FRASTM commented:
Unfortunately, the https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/flash/flash_mcux_flexspi_nor.c is using the macro with **decl_nph =1**
 static int flash_flexspi_nor_probe(struct flash_flexspi_nor_data *data)
```
{
	uint32_t flexspi_lut[FLEXSPI_INSTR_END][MEMC_FLEXSPI_CMD_PER_SEQ] = {0};
	/* JESD216B defines up to 23 basic flash parameters */
	uint32_t param_buf[23];
	/* Space to store SFDP header and first parameter header */
	uint8_t sfdp_buf[JESD216_SFDP_SIZE(1)] __aligned(4)
```;

FRASTM commented:
A proposal could be to define two macro
```
#define JESD216_SFDP_SIZE_1 (sizeof(struct jesd216_sfdp_header)	\
				+ ((1) * sizeof(struct jesd216_param_header)))
```

```
#define JESD216_SFDP_SIZE (sizeof(struct jesd216_sfdp_header)	\
				+ ((2) * sizeof(struct jesd216_param_header)))
```

erwango commented:
> A proposal could be to define two macro
> 
> ```
> #define JESD216_SFDP_SIZE_1 (sizeof(struct jesd216_sfdp_header)	\
> 				+ ((1) * sizeof(struct jesd216_param_header)))
> ```
> 
> ```
> #define JESD216_SFDP_SIZE_2 (sizeof(struct jesd216_sfdp_header)	\
> 				+ ((2) * sizeof(struct jesd216_param_header)))
> ```

@de-nordic Is that fine to you ?

erwango commented:
> @de-nordic Is that fine to you ?

Gentle ping @de-nordic

erwango commented:
@de-nordic Just for the sake of the discussion. There was an alternative proposal [here](https://github.com/zephyrproject-rtos/zephyr/pull/65813#issuecomment-1929874828) aim was to have only one version of the macro (ie not depending on MISRA enabled or not).
Do you prefer current state of the PR vs this alternative ?

de-nordic commented:
> @de-nordic Just for the sake of the discussion. There was an alternative proposal [here](https://github.com/zephyrproject-rtos/zephyr/pull/65813#issuecomment-1929874828) aim was to have only one version of the macro (ie not depending on MISRA enabled or not). Do you prefer current state of the PR vs this alternative ?

I will answer this with a dumb question: wouldn't replacing the definitions of `const uint8_t decl_nph = 2;` with preprocessor define on top of unit, as affected unions are locally defined, fix the issue completely and we could leave the macro as is?
